File: problems/biggesttriangle/submissions/accepted/biggesttriangle-zf.py
#!/usr/bin/python3
EPS = 1e-8

def cross(p, q):
    return (p*q.conjugate()).imag

def isect(l1, l2):
    den = cross(l1[1]-l1[0], l2[0]-l2[1])
    if abs(den) < EPS: return None

    num = cross(l2[0]-l1[0], l2[0]-l2[1])

    return l1[0]+(l1[1]-l1[0])*num/den

def area(l1, l2, l3):
    i1 = isect(l1, l2)
    i2 = isect(l2, l3)
    i3 = isect(l3, l1)

    if None in [i1, i2, i3]: return 0.0

    return abs(i1-i2) + abs(i2-i3) + abs(i3-i1)

n = int(input())
lines = []

for _ in range(n):
    x1, y1, x2, y2 = map(float, input().split())
    lines.append([complex(x1, y1), complex(x2, y2)])

best = 0.0
for i in range(n):
    for j in range(i):
        for k in range(j):
            best = max(best, area(lines[i], lines[j], lines[k]))

if abs(best) < EPS:
    print("no triangle")
else:
    print("{0:.10f}".format(best))

File: problems/biggesttriangle/submissions/accepted/biggesttriangle-zf.cpp
#include <iostream>
#include <algorithm>
#include <complex>
#include <vector>
#include <iomanip>

#define EPS 1e-8

using namespace std;

typedef double ld;

typedef complex<ld> point;

struct line {
  point p, q;
};

ld cross(point p, point q) {
  return imag(p*conj(q));
}

bool isect_iline(line l1, line l2, point& isect) {
  ld den = cross(l1.q-l1.p, l2.p-l2.q);

  if (fabs(den) < EPS) return false;

  ld num = cross(l2.p-l1.p, l2.p-l2.q);

  isect = l1.p + (l1.q-l1.p)*num/den;

  return true;
}

ld area(line l1, line l2, line l3) {
  point i1, i2, i3;

  if (!isect_iline(l1, l2, i1)) return 0;
  if (!isect_iline(l2, l3, i2)) return 0;
  if (!isect_iline(l3, l1, i3)) return 0;

  return abs(i1-i2) + abs(i2-i3) + abs(i3-i1);
}

int main() {
  int n;
  cin >> n;

  vector<line> lines(n);
  for (auto& l : lines) {
    ld x1, y1, x2, y2;
    cin >> x1 >> y1 >> x2 >> y2;
    l.p = point(x1, y1);
    l.q = point(x2, y2);
  }

  ld best = 0;
  for (int i = 0; i < n; ++i)
    for (int j = 0; j < i; ++j)
      for (int k = 0; k < j; ++k)
        best = max(best, area(lines[i], lines[j], lines[k]));

  cout.setf(ios::fixed);
  cout.precision(10);

  if (best < EPS) cout << "no triangle" << endl;
  else cout << best << endl;

  return 0;
}
